Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
  * limitations under the License.
  */
 
-// FIXME: placeholder.
-public typealias Metadata = [String: String]
+import Foundation
+
+/// A collection of metadata key-value pairs, found in RPC streams.
+/// A key can have multiple values associated to it.
+/// Values can be either strings or binary data, in the form of `Data`.
+/// - Note: Binary values must have keys ending in `-bin`, and this will be checked when adding pairs.
+public struct Metadata: Sendable, Hashable {
+
+  /// A metadata value. It can either be a simple string, or binary data.
+  public enum MetadataValue: Sendable, Hashable {
+    case string(String)
+    case binary(Data)
+  }
+
+  /// A metadata key-value pair.
+  public struct MetadataKeyValue: Sendable, Hashable {
+    internal let key: String
+    internal let value: MetadataValue
+
+    /// Constructor for a metadata key-value pair.
+    /// - Parameters:
+    ///   - key: The key for the key-value pair.
+    ///   - value: The value to be associated to the given key. If it's a binary value, then the associated
+    ///   key must end in `-bin`, otherwise, this method will produce an assertion failure.
+    init(key: String, value: MetadataValue) {
+      if case .binary = value {
+        assert(key.hasSuffix("-bin"), "Keys for binary values must end in -bin")
+      }
+      self.key = key
+      self.value = value
+    }
+  }
+
+  private let lockedElements: LockedValueBox<[MetadataKeyValue]>
+  private var elements: [MetadataKeyValue] {
+    get {
+      self.lockedElements.withLockedValue { $0 }
+    }
+    set {
+      self.lockedElements.withLockedValue { $0 = newValue }
+    }
+  }
+
+  /// The Metadata collection's capacity.
```

Comment:
Presumably `count` and `isEmpty` come for free with `RandomAccessCollection` conformance?